// lerna-changelog/lib/progress-bar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ProgressBar = require("progress");
var padEnd = require("string.prototype.padend");
var ProgressBarController = (function () {
    function ProgressBarController() {
        this.bar = null;
    }
    ProgressBarController.prototype.init = function (total) {
        if (this.bar) {
            this.terminate();
        }
        if (!process.stdout.isTTY) {
            return;
        }
        this.bar = new ProgressBar(":packagename ╢:bar╟", {
            total: total,
            complete: "█",
            incomplete: "░",
            clear: true,
            width: (process.stdout.columns || 100) - 50 - 3
        });
    };
    ProgressBarController.prototype.setTitle = function (name) {
        if (this.bar) {
            this.bar.tick(0, {
                packagename: padEnd(name.slice(0, 50), 50)
            });
        }
    };
    ProgressBarController.prototype.tick = function () {
        if (this.bar) {
            this.bar.tick(1);
        }
    };
    ProgressBarController.prototype.clear = function () {
        if (this.bar) {
            this.bar.terminate();
        }
    };
    ProgressBarController.prototype.terminate = function () {
        this.clear();
        this.bar = null;
    };
    return ProgressBarController;
}());
exports.default = new ProgressBarController();
